src/main/scala/mesosphere/marathon/storage/migration/Migration.scala

```diff
@@ -119,32 +119,28 @@ class Migration(
     // step 2: restore state from given backup
     await(config.restore.map(uri => backup.restore(new URI(uri))).getOrElse(Future.successful(Done)))
 
-    // mark migration as started
-    await(persistenceStore.startMigration())
-
-    // run the migration, to ensure we can operate on the zk state
-    await(migrateStorage(backupCreated = config.backup.isDefined || config.restore.isDefined).asTry) match {
-      case Success(result) =>
-        // mark migration as completed
-        await(persistenceStore.endMigration())
-
-        logger.info(s"Migration successfully applied for version ${StorageVersions.current.str}")
-        result
-      case Failure(ex: MigrationCancelledException) =>
-        logger.error(ex.getMessage)
-
-        // mark migration as completed
-        await(persistenceStore.endMigration())
-
-        throw new MigrationFailedException("Migration cancelled", ex.getCause)
-      case Failure(ex) =>
-        throw ex
-    }
+    // last step: run the migration, to ensure we can operate on the zk state
+    val result = await(migrateStorage(backupCreated = config.backup.isDefined || config.restore.isDefined))
+    logger.info(s"Migration successfully applied for version ${StorageVersions.current.str}")
+    result
   }
 
   def migrate(): Seq[StorageVersion] =
     Await.result(migrateAsync(), Duration.Inf)
 
+  @SuppressWarnings(Array("all")) // async/await
+  def runMigrations(version: Protos.StorageVersion, backupCreated: Boolean = false): Future[Seq[StorageVersion]] =
+    async {
+      if (!backupCreated && config.backupLocation.isDefined) {
+        logger.info("Backup current state")
+        await(backup.backup(config.backupLocation.get))
+        logger.info("Backup finished. Apply migration.")
```

Comment:
I'd definitely add where the backup was created e.g. `logger.info("Backup successfully saved to ${config.backupLocation}...`.

Reply:
Done.